drivers/serial/uart_nrfx_uarte.c

```diff
@@ -251,6 +323,472 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+
+	size_t rx_len = nrf_uarte_rx_amount_get(uarte)
+			- data->async->rx_offset;
+	struct uart_event evt = {
+		.type = UART_RX_RDY,
+		.data.rx.buf = data->async->rx_buf,
+		.data.rx.len = rx_len,
+		.data.rx.offset = data->async->rx_offset,
+	};
+	user_callback(dev, &evt);
+
+	evt.type = UART_RX_BUF_RELEASED;
+	evt.data.rx_buf.buf = data->async->rx_buf;
+	user_callback(dev, &evt);
+
+	if (data->async->rx_next_buf) {
+		data->async->rx_buf = data->async->rx_next_buf;
+		data->async->rx_next_buf = NULL;
+
+		data->async->rx_total_byte_cnt += rx_len;
+		data->async->rx_total_user_byte_cnt =
+			data->async->rx_total_byte_cnt;
+		data->async->rx_offset = 0;
+	} else {
+		data->async->rx_buf = NULL;
+		evt.type = UART_RX_DISABLED;
+		user_callback(dev, &evt);
+	}
+}
+
+/* This handler is called when the reception is interrupted, in contrary to
+ * finishing the reception after filling all provided buffers, in which case
+ * the events UART_RX_BUF_RELEASED and UART_RX_DISABLED are reported
+ * from endrx_isr.
+ */
+static void rxto_isr(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	struct uart_event evt = {
+		.type = UART_RX_BUF_RELEASED,
+		.data.rx_buf.buf = data->async->rx_buf,
+	};
+	user_callback(dev, &evt);
+	data->async->rx_buf = NULL;
+	if (data->async->rx_next_buf) {
+		evt.type = UART_RX_BUF_RELEASED;
+		evt.data.rx_buf.buf = data->async->rx_next_buf;
+		user_callback(dev, &evt);
+		data->async->rx_next_buf = NULL;
+	}
+	evt.type = UART_RX_DISABLED;
+	user_callback(dev, &evt);
+}
+
+static void txstopped_isr(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	if (!data->async->tx_buf) {
+		return;
+	}
+
+	size_t amount = nrf_uarte_tx_amount_get(get_uarte_instance(dev));
+
+	struct uart_event evt = {
+		.data.tx.buf = data->async->tx_buf,
+		.data.tx.len = amount,
+	};
+	if (amount == data->async->tx_size) {
+		evt.type = UART_TX_DONE;
+	} else {
+		evt.type = UART_TX_ABORTED;
+	}
+	data->async->tx_buf = NULL;
+	data->async->tx_size = 0;
+	user_callback(dev, &evt);
+}
+
+static void endtx_isr(struct device *dev)
+{
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
+	k_timer_stop(&data->async->tx_timeout_timer);
+}
+
+static void uarte_nrfx_isr(struct device *dev)
+{
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	if (!hw_rx_counting_enabled(data)
+	    && nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_RXDRDY)) {
+		nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_RXDRDY);
+		data->async->rx_cnt.cnt++;
+		return;
+	}
+
+	if (nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ERROR)) {
+		nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ERROR);
+		error_isr(dev);
+	}
+
+	if (nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_RXSTARTED)) {
+		nrf_uarte_event_clear(uarte,
+				      NRF_UARTE_EVENT_RXSTARTED);
```

Comment:
Nit: this line doesn't seem to need this break.
Same story below, with `NRF_UARTE_EVENT_TXSTOPPED`.